bauble/src/convert.rs

```diff
@@ -0,0 +1,359 @@
+                                .ok_or_else(|| DeserializeError::Custom {
+                                    message: format!("{} is not a valid {}", number, stringify!($ty)),
+                                    span,
+                                })?;
+                            // SAFETY: No allocations are contained in these types.
+                            unsafe { allocator.wrap(number) }
+                        }
+                    )
+                }
+            }
+        )*
+    };
+}
+
+impl_nums! {
+    u8, u16, u32, u64, u128, usize,
+    i8, i16, i32, i64, i128, isize,
+    f32, f64,
+}
+
+impl<'a, A: BaubleAllocator<'a>> FromBauble<'a, A> for bool {
+    fn from_bauble(val: Val, allocator: &'a A) -> Result<A::Out<Self>, Box<DeserializeError>> {
+        match_val!(
+            val,
+            (Bool(v), _span) => {
+                // SAFETY: No allocations are contained in a bool.
+                unsafe { allocator.wrap(v) }
+            }
+        )
+    }
+}
+
+impl<'a> FromBauble<'a> for String {
+    fn from_bauble(val: Val, _: &DefaultAllocator) -> Result<Self, Box<DeserializeError>> {
+        match_val!(
+            val,
+            (Str(string), _span) => string,
+        )
+    }
+}
+
+impl<'a, A: BaubleAllocator<'a>, T: FromBauble<'a, A>> FromBauble<'a, A> for Option<T> {
+    fn from_bauble(val: Val, allocator: &'a A) -> Result<A::Out<Option<T>>, Box<DeserializeError>> {
+        match_val!(
+            val,
+            (Opt(opt), _span) => {
+                let opt = opt
+                .map(|val| T::from_bauble(*val, allocator).and_then(|t| {
+                    // SAFETY: We wrap this value again inside the option.
+                    unsafe { allocator.validate(t) }
+                }))
+                .transpose();
+
+                // SAFETY: The contained value in the option was validated,
+                // and outside of that option doesn't contain any allocations.
+                unsafe { opt.map(|opt| allocator.wrap(opt))? }
+            }
+        )
+    }
+}
+
+impl<'a> FromBauble<'a> for () {
+    fn from_bauble(val: Val, _: &DefaultAllocator) -> Result<Self, Box<DeserializeError>> {
+        match_val!(
+            val,
+            (Unit, _span) => (),
+        )
+    }
+}
+
+macro_rules! impl_tuple {
+    ($($ident:ident),+) => {
+        impl<'a, $($ident: FromBauble<'a>),*> FromBauble<'a> for ($($ident),*,) {
+            fn from_bauble(
+                val: Val,
+                allocator: &'a DefaultAllocator,
+            ) -> Result<Self, Box<DeserializeError>> {
+                const LEN: usize = [$(stringify!($ident)),*].len();
+                match_val!(
+                    val,
+                    (Tuple(seq), span) => {
+                        if seq.len() == LEN {
+                            let mut seq = seq.into_iter();
+                            ($($ident::from_bauble(seq.next().expect("We checked tuple length"), allocator)?),*,)
```

Comment:
This could probably be an `unwrap_unchecked()`

Reply:
Yeah that works, added that with a safety comment